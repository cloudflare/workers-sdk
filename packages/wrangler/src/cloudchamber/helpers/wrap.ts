export async function wrap<T, E extends Error>(
	fn: Promise<T>
): Promise<[T, null] | [null, E]> {
	return fn
		.then((data) => [data, null] as [T, null])
		.catch((err) => [null, err as unknown as E] as [null, E]);
}
