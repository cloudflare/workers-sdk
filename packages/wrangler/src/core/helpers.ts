import { CommandLineArgsError } from "../errors";
import type { CreateCommandResult } from "./create-command";
import type {
	AliasDefinition,
	CommandDefinition,
	NamedArgDefinitions,
	NamespaceDefinition,
} from "./types";

/**
 * A helper to demand one of a set of options
 * via https://github.com/yargs/yargs/issues/1093#issuecomment-491299261
 */
export function demandOneOfOption(...options: string[]) {
	return function (argv: { [key: string]: unknown }) {
		const count = options.filter((option) => argv[option]).length;
		const lastOption = options.pop();

		if (count === 0) {
			throw new CommandLineArgsError(
				`Exactly one of the arguments ${options.join(
					", "
				)} and ${lastOption} is required`
			);
		} else if (count > 1) {
			throw new CommandLineArgsError(
				`Arguments ${options.join(
					", "
				)} and ${lastOption} are mutually exclusive`
			);
		}

		return true;
	};
}

/**
 * A helper to ensure that an argument only receives a single value.
 *
 * This is a workaround for a limitation in yargs where non-array arguments can still receive multiple values
 * even though the inferred type is not an array.
 *
 * @see https://github.com/yargs/yargs/issues/1318
 */
export function demandSingleValue<Argv extends { [key: string]: unknown }>(
	key: string,
	allow?: (argv: Argv) => boolean
) {
	return function (argv: Argv) {
		if (Array.isArray(argv[key]) && !allow?.(argv)) {
			throw new CommandLineArgsError(
				`The argument "--${key}" expects a single value, but received multiple: ${JSON.stringify(argv[key])}.`
			);
		}

		return true;
	};
}

/**
 * Checks if a definition is an alias definition.
 */
export function isAliasDefinition(
	def:
		| AliasDefinition
		| CreateCommandResult<NamedArgDefinitions>
		| NamespaceDefinition
): def is AliasDefinition {
	return (def as AliasDefinition).aliasOf !== undefined;
}

/**
 * Checks if a definition is a command definition.
 */
export function isCommandDefinition(
	def:
		| AliasDefinition
		| CreateCommandResult<NamedArgDefinitions>
		| NamespaceDefinition
): def is CommandDefinition {
	return (def as CommandDefinition).handler !== undefined;
}

/**
 * Checks if a definition is a namespace definition.
 */
export function isNamespaceDefinition(
	def:
		| AliasDefinition
		| CreateCommandResult<NamedArgDefinitions>
		| NamespaceDefinition
): def is NamespaceDefinition {
	return !isAliasDefinition(def) && !isCommandDefinition(def);
}
