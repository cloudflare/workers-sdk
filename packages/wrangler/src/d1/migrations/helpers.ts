import fs from "node:fs";
import path from "path";
import { confirm } from "../../dialogs";
import { CI } from "../../is-ci";
import isInteractive from "../../is-interactive";
import { logger } from "../../logger";
import { DEFAULT_MIGRATION_PATH } from "../constants";
import { executeSql } from "../execute";
import type { ConfigFields, DevConfig, Environment } from "../../config";
import type { QueryResult } from "../execute";
import type { Migration } from "../types";

export async function getMigrationsPath({
	projectPath,
	migrationsFolderPath,
	createIfMissing,
}: {
	projectPath: string;
	migrationsFolderPath: string;
	createIfMissing: boolean;
}): Promise<string> {
	const dir = path.resolve(projectPath, migrationsFolderPath);
	if (fs.existsSync(dir)) return dir;

	const warning = `No migrations folder found.${
		migrationsFolderPath === DEFAULT_MIGRATION_PATH
			? " Set `migrations_dir` in wrangler.toml to choose a different path."
			: ""
	}`;

	if (createIfMissing && (await confirm(`${warning}\nOk to create ${dir}?`))) {
		fs.mkdirSync(dir, { recursive: true });
		return dir;
	} else {
		logger.warn(warning);
	}

	throw new Error(`No migrations present at ${dir}.`);
}

export async function getUnappliedMigrations({
	migrationsTableName,
	migrationsPath,
	local,
	config,
	name,
	persistTo,
	preview,
}: {
	migrationsTableName: string;
	migrationsPath: string;
	local: boolean | undefined;
	config: ConfigFields<DevConfig> & Environment;
	name: string;
	persistTo: string | undefined;
	preview: boolean | undefined;
}): Promise<Array<string>> {
	const appliedMigrations = (
		await listAppliedMigrations(
			migrationsTableName,
			local,
			config,
			name,
			persistTo,
			preview
		)
	).map((migration) => {
		return migration.name;
	});
	const projectMigrations = getMigrationNames(migrationsPath);

	const unappliedMigrations: Array<string> = [];

	for (const migration of projectMigrations) {
		if (!appliedMigrations.includes(migration)) {
			unappliedMigrations.push(migration);
		}
	}

	return unappliedMigrations;
}

const listAppliedMigrations = async (
	migrationsTableName: string,
	local: boolean | undefined,
	config: ConfigFields<DevConfig> & Environment,
	name: string,
	persistTo: string | undefined,
	preview: boolean | undefined
): Promise<Migration[]> => {
	const response: QueryResult[] | null = await executeSql({
		local,
		config,
		name,
		shouldPrompt: isInteractive() && !CI.isCI(),
		persistTo,
		command: `SELECT *
		FROM ${migrationsTableName}
		ORDER BY id`,
		file: undefined,
		json: undefined,
		preview,
	});

	if (!response || response[0].results.length === 0) return [];

	return response[0].results as Migration[];
};

function getMigrationNames(migrationsPath: string): Array<string> {
	const migrations = [];

	const dir = fs.opendirSync(migrationsPath);

	let dirent;
	while ((dirent = dir.readSync()) !== null) {
		if (dirent.name.endsWith(".sql")) migrations.push(dirent.name);
	}

	dir.closeSync();

	return migrations;
}

export function getNextMigrationNumber(migrationsPath: string): number {
	let highestMigrationNumber = -1;

	for (const migration in getMigrationNames(migrationsPath)) {
		const migrationNumber = parseInt(migration.split("_")[0]);

		if (migrationNumber > highestMigrationNumber) {
			highestMigrationNumber = migrationNumber;
		}
	}

	return highestMigrationNumber + 1;
}

export const initMigrationsTable = async ({
	migrationsTableName,
	local,
	config,
	name,
	persistTo,
	preview,
}: {
	migrationsTableName: string;
	local: boolean | undefined;
	config: ConfigFields<DevConfig> & Environment;
	name: string;
	persistTo: string | undefined;
	preview: boolean | undefined;
}) => {
	return executeSql({
		local,
		config,
		name,
		shouldPrompt: isInteractive() && !CI.isCI(),
		persistTo,
		command: `CREATE TABLE IF NOT EXISTS ${migrationsTableName}(
		id         INTEGER PRIMARY KEY AUTOINCREMENT,
		name       TEXT UNIQUE,
		applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL
);`,
		file: undefined,
		json: undefined,
		preview,
	});
};
