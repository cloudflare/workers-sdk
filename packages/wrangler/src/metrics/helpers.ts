import os from "node:os";
import { version as wranglerVersion } from "../../package.json";

export function getWranglerVersion() {
	return wranglerVersion;
}
// used by "new" metrics
export function getPlatform() {
	const platform = os.platform();

	switch (platform) {
		case "win32":
			return "Windows";
		case "darwin":
			return "Mac OS";
		case "linux":
			return "Linux";
		default:
			return `Others: ${platform}`;
	}
}

// used by "old" metrics
export function getOS() {
	return process.platform + ":" + process.arch;
}

export function getOSVersion() {
	return os.version();
}

export function getNodeVersion() {
	const nodeVersion = process.versions.node;
	return parseInt(nodeVersion.split(".")[0]);
}
