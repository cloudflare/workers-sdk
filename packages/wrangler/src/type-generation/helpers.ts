import { readFileSync } from "fs";
import { version } from "workerd";
import { logger } from "../logger";
import { generateEnvTypes } from ".";
import type { Config } from "../config";
import type { Entry } from "../deployment-bundle/entry";

// Checks the default location for a generated types file and compares if the
// recorded Env hash, workerd version or compat date and flags have changed
// compared to the current values in the config. Prompts user to re-run wrangler
// types if so.
export const checkTypesDiff = async (config: Config, entry: Entry) => {
	if (!entry.file.endsWith(".ts")) {
		return;
	}
	let maybeExistingTypesFileLines: string[];
	try {
		// Checking the default location only
		maybeExistingTypesFileLines = readFileSync(
			"./worker-configuration.d.ts",
			"utf-8"
		).split("\n");
	} catch {
		return;
	}
	const existingEnvHeader = maybeExistingTypesFileLines.find((line) =>
		line.startsWith("// Generated by Wrangler by running")
	);
	const maybeExistingHash =
		existingEnvHeader?.match(/hash: (?<hash>.*)\)/)?.groups?.hash;
	const previousStrictVars = existingEnvHeader?.match(
		/--strict-vars(=|\s)(?<result>true|false)/
	)?.groups?.result;
	const previousEnvInterface = existingEnvHeader?.match(
		/--env-interface(=|\s)(?<result>[a-zA-Z][a-zA-Z0-9_]*)/
	)?.groups?.result;

	let newEnvHeader: string | undefined;
	try {
		const { envHeader } = await generateEnvTypes(
			config,
			{ strictVars: previousStrictVars === "false" ? false : true },
			previousEnvInterface ?? "Env",
			"worker-configuration.d.ts",
			entry,
			// don't log anything
			false
		);
		newEnvHeader = envHeader;
	} catch (e) {
		logger.error(e);
	}

	const newHash = newEnvHeader?.match(/hash: (?<hash>.*)\)/)?.groups?.hash;

	const existingRuntimeHeader = maybeExistingTypesFileLines.find((line) =>
		line.startsWith("// Runtime types generated with")
	);
	const newRuntimeHeader = `// Runtime types generated with workerd@${version} ${config.compatibility_date} ${config.compatibility_flags.sort().join(",")}`;

	const envOutOfDate = existingEnvHeader && maybeExistingHash !== newHash;
	const runtimeOutOfDate =
		existingRuntimeHeader && existingRuntimeHeader !== newRuntimeHeader;

	if (envOutOfDate || runtimeOutOfDate) {
		logger.log(
			"‚ùì It looks like your types might be out of date. Have you updated your config file since last running `wrangler types`?"
		);
	}
};
