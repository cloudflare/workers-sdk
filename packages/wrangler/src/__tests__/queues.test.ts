import { rest } from "msw";
import { mockAccountId, mockApiToken } from "./helpers/mock-account-id";
import { mockConsoleMethods } from "./helpers/mock-console";
import { msw } from "./helpers/msw";
import { runInTempDir } from "./helpers/run-in-tmp";
import { runWrangler } from "./helpers/run-wrangler";
import type { QueueResponse, PostConsumerBody } from "../queues/client";

describe("wrangler", () => {
	mockAccountId();
	mockApiToken();
	runInTempDir();
	const std = mockConsoleMethods();

	describe("queues", () => {
		it("should show the correct help text", async () => {
			await runWrangler("queues --help");
			expect(std.err).toMatchInlineSnapshot(`""`);
			expect(std.out).toMatchInlineSnapshot(`
			"wrangler queues

			ðŸ‡¶ Configure Workers Queues

			Commands:
			  wrangler queues list           List Queues
			  wrangler queues create <name>  Create a Queue
			  wrangler queues delete <name>  Delete a Queue
			  wrangler queues consumer       Configure Queue Consumers

			Flags:
			  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
			  -c, --config                    Path to .toml configuration file  [string]
			  -e, --env                       Environment to use for operations and .env files  [string]
			  -h, --help                      Show help  [boolean]
			  -v, --version                   Show version number  [boolean]"
		`);
		});

		describe("list", () => {
			function mockListRequest(queues: QueueResponse[], page: number) {
				const requests = { count: 0 };
				msw.use(
					rest.get(
						"*/accounts/:accountId/workers/queues?*",
						async (request, response, context) => {
							requests.count += 1;
							const query = request.url.searchParams;
							expect(Number(query.get("page"))).toEqual(page);
							expect(await request.text()).toEqual("");
							return response.once(
								context.json({
									success: true,
									errors: [],
									messages: [],
									result: queues,
								})
							);
						}
					)
				);
				return requests;
			}

			it("should show the correct help text", async () => {
				await runWrangler("queues list --help");
				expect(std.err).toMatchInlineSnapshot(`""`);
				expect(std.out).toMatchInlineSnapshot(`
			"wrangler queues list

			List Queues

			Flags:
			  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
			  -c, --config                    Path to .toml configuration file  [string]
			  -e, --env                       Environment to use for operations and .env files  [string]
			  -h, --help                      Show help  [boolean]
			  -v, --version                   Show version number  [boolean]

			Options:
			      --page  Page number for pagination  [number]"
		`);
			});

			it("should list queues on page 1 with no --page", async () => {
				const expectedQueues: QueueResponse[] = [
					{
						queue_name: "queue-1",
						created_on: "01-01-2001",
						modified_on: "01-01-2001",
					},
					{
						queue_name: "queue-2",
						created_on: "01-01-2001",
						modified_on: "01-01-2001",
					},
				];
				const expectedPage = 1;
				mockListRequest(expectedQueues, expectedPage);
				await runWrangler("queues list");

				expect(std.err).toMatchInlineSnapshot(`""`);
				const buckets = JSON.parse(std.out);
				expect(buckets).toEqual(expectedQueues);
			});

			it("should list queues using --page=2", async () => {
				const expectedQueues: QueueResponse[] = [
					{
						queue_name: "queue-100",
						created_on: "01-01-2001",
						modified_on: "01-01-2001",
					},
				];
				const expectedPage = 2;
				mockListRequest(expectedQueues, expectedPage);
				await runWrangler("queues list --page=2");

				expect(std.err).toMatchInlineSnapshot(`""`);
				const buckets = JSON.parse(std.out);
				expect(buckets).toEqual(expectedQueues);
			});
		});

		describe("create", () => {
			function mockCreateRequest(expectedQueueName: string) {
				const requests = { count: 0 };

				msw.use(
					rest.post(
						"*/accounts/:accountId/workers/queues",
						async (request, response, context) => {
							requests.count += 1;
							const body = (await request.json()) as {
								queue_name: string;
							};
							expect(body.queue_name).toEqual(expectedQueueName);
							return response.once(
								context.json({
									success: true,
									errors: [],
									messages: [],
									result: {
										queue_name: expectedQueueName,
										created_on: "01-01-2001",
										modified_on: "01-01-2001",
									},
								})
							);
						}
					)
				);
				return requests;
			}

			it("should show the correct help text", async () => {
				await runWrangler("queues create --help");
				expect(std.err).toMatchInlineSnapshot(`""`);
				expect(std.out).toMatchInlineSnapshot(`
			"wrangler queues create <name>

			Create a Queue

			Positionals:
			  name  The name of the queue  [string] [required]

			Flags:
			  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
			  -c, --config                    Path to .toml configuration file  [string]
			  -e, --env                       Environment to use for operations and .env files  [string]
			  -h, --help                      Show help  [boolean]
			  -v, --version                   Show version number  [boolean]"
		`);
			});

			it("should create a queue", async () => {
				const requests = mockCreateRequest("testQueue");
				await runWrangler("queues create testQueue");
				expect(std.out).toMatchInlineSnapshot(`
					"Creating queue testQueue.
					Created queue testQueue."
			  `);
				expect(requests.count).toEqual(1);
			});

			it("should show link to dash when not enabled", async () => {
				const queueName = "testQueue";
				msw.use(
					rest.post(
						"*/accounts/:accountId/workers/queues",
						async (request, response, context) => {
							expect(request.params.accountId).toEqual("some-account-id");
							return response.once(
								context.status(403),
								context.json({
									success: false,
									errors: [
										{ message: "workers.api.error.unauthorized", code: 10023 },
									],
									messages: [],
								})
							);
						}
					)
				);

				await expect(
					runWrangler(`queues create ${queueName}`)
				).rejects.toThrowError();
				expect(std.out).toMatchInlineSnapshot(`
			"Creating queue testQueue.
			Queues is not currently enabled on this account. Go to https://dash.cloudflare.com/some-account-id/workers/queues to enable it.

			[31mX [41;31m[[41;97mERROR[41;31m][0m [1mA request to the Cloudflare API (/accounts/some-account-id/workers/queues) failed.[0m

			  workers.api.error.unauthorized [code: 10023]

			  If you think this is a bug, please open an issue at:
			  [4mhttps://github.com/cloudflare/workers-sdk/issues/new/choose[0m

			"
		`);
			});
		});

		describe("delete", () => {
			function mockDeleteRequest(expectedQueueName: string) {
				const requests = { count: 0 };
				msw.use(
					rest.delete(
						"*/accounts/:accountId/workers/queues/:queueName",
						async (request, response, context) => {
							requests.count += 1;
							expect(request.params.queueName).toEqual(expectedQueueName);
							expect(request.params.accountId).toEqual("some-account-id");
							return response.once(
								context.json({
									success: true,
									errors: [],
									messages: [],
									result: {},
								})
							);
						}
					)
				);
				return requests;
			}

			it("should show the correct help text", async () => {
				await runWrangler("queues delete --help");
				expect(std.err).toMatchInlineSnapshot(`""`);
				expect(std.out).toMatchInlineSnapshot(`
			"wrangler queues delete <name>

			Delete a Queue

			Positionals:
			  name  The name of the queue  [string] [required]

			Flags:
			  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
			  -c, --config                    Path to .toml configuration file  [string]
			  -e, --env                       Environment to use for operations and .env files  [string]
			  -h, --help                      Show help  [boolean]
			  -v, --version                   Show version number  [boolean]"
		`);
			});

			it("should delete a queue", async () => {
				const requests = mockDeleteRequest("testQueue");
				await runWrangler("queues delete testQueue");
				expect(std.out).toMatchInlineSnapshot(`
					"Deleting queue testQueue.
					Deleted queue testQueue."
			  `);
				expect(requests.count).toEqual(1);
			});
		});

		describe("consumers", () => {
			it("should show the correct help text", async () => {
				await runWrangler("queues consumer --help");

				expect(std.err).toMatchInlineSnapshot(`""`);
				expect(std.out).toMatchInlineSnapshot(`
			"wrangler queues consumer

			Configure Queue Consumers

			Commands:
			  wrangler queues consumer add <queue-name> <script-name>     Add a Queue Consumer
			  wrangler queues consumer remove <queue-name> <script-name>  Remove a Queue Consumer

			Flags:
			  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
			  -c, --config                    Path to .toml configuration file  [string]
			  -e, --env                       Environment to use for operations and .env files  [string]
			  -h, --help                      Show help  [boolean]
			  -v, --version                   Show version number  [boolean]"
		`);
			});

			describe("add", () => {
				function mockPostRequest(
					expectedQueueName: string,
					expectedBody: PostConsumerBody
				) {
					const requests = { count: 0 };
					msw.use(
						rest.post(
							"*/accounts/:accountId/workers/queues/:queueName/consumers",
							async (request, response, context) => {
								requests.count += 1;
								expect(request.params.queueName).toEqual(expectedQueueName);
								expect(request.params.accountId).toEqual("some-account-id");
								expect(await request.json()).toEqual(expectedBody);
								return response.once(
									context.json({
										success: true,
										errors: [],
										messages: [],
										result: {},
									})
								);
							}
						)
					);
					return requests;
				}

				it("should show the correct help text", async () => {
					await runWrangler("queues consumer add --help");
					expect(std.err).toMatchInlineSnapshot(`""`);
					expect(std.out).toMatchInlineSnapshot(`
				"wrangler queues consumer add <queue-name> <script-name>

				Add a Queue Consumer

				Positionals:
				  queue-name   Name of the queue to configure  [string] [required]
				  script-name  Name of the consumer script  [string] [required]

				Flags:
				  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
				  -c, --config                    Path to .toml configuration file  [string]
				  -e, --env                       Environment to use for operations and .env files  [string]
				  -h, --help                      Show help  [boolean]
				  -v, --version                   Show version number  [boolean]

				Options:
				      --batch-size         Maximum number of messages per batch  [number]
				      --batch-timeout      Maximum number of seconds to wait to fill a batch with messages  [number]
				      --message-retries    Maximum number of retries for each message  [number]
				      --dead-letter-queue  Queue to send messages that failed to be consumed  [string]
				      --max-concurrency    The maximum number of concurrent consumer Worker invocations. Must be a positive integer  [number]"
			`);
				});

				it("should add a consumer using defaults", async () => {
					const expectedBody: PostConsumerBody = {
						script_name: "testScript",
						environment_name: "",
						settings: {
							batch_size: undefined,
							max_retries: undefined,
							max_wait_time_ms: undefined,
							max_concurrency: undefined,
						},
						dead_letter_queue: undefined,
					};
					mockPostRequest("testQueue", expectedBody);
					await runWrangler("queues consumer add testQueue testScript");
					expect(std.out).toMatchInlineSnapshot(`
							"Adding consumer to queue testQueue.
							Added consumer to queue testQueue."
					`);
				});

				it("should add a consumer using custom values", async () => {
					const expectedBody: PostConsumerBody = {
						script_name: "testScript",
						environment_name: "myEnv",
						settings: {
							batch_size: 20,
							max_retries: 3,
							max_wait_time_ms: 10 * 1000,
							max_concurrency: 3,
						},
						dead_letter_queue: "myDLQ",
					};
					mockPostRequest("testQueue", expectedBody);

					await runWrangler(
						"queues consumer add testQueue testScript --env myEnv --batch-size 20 --batch-timeout 10 --message-retries 3 --max-concurrency 3 --dead-letter-queue myDLQ"
					);
					expect(std.out).toMatchInlineSnapshot(`
						"Adding consumer to queue testQueue.
						Added consumer to queue testQueue."
					`);
				});

				it("should show link to dash when not enabled", async () => {
					const queueName = "testQueue";
					msw.use(
						rest.post(
							"*/accounts/:accountId/workers/queues/:queueName/consumers",
							async (request, response, context) => {
								expect(request.params.queueName).toEqual(queueName);
								expect(request.params.accountId).toEqual("some-account-id");
								return response.once(
									context.status(403),
									context.json({
										success: false,
										errors: [
											{
												code: 10023,
												message: "workers.api.error.unauthorized",
											},
										],
										messages: [],
										result: {},
									})
								);
							}
						)
					);

					await expect(
						runWrangler(`queues consumer add ${queueName} testScript`)
					).rejects.toThrowError();
					expect(std.out).toMatchInlineSnapshot(`
				"Adding consumer to queue testQueue.
				Queues is not currently enabled on this account. Go to https://dash.cloudflare.com/some-account-id/workers/queues to enable it.

				[31mX [41;31m[[41;97mERROR[41;31m][0m [1mA request to the Cloudflare API (/accounts/some-account-id/workers/queues/testQueue/consumers) failed.[0m

				  workers.api.error.unauthorized [code: 10023]

				  If you think this is a bug, please open an issue at:
				  [4mhttps://github.com/cloudflare/workers-sdk/issues/new/choose[0m

				"
			`);
				});
			});

			describe("delete", () => {
				function mockDeleteRequest(
					expectedQueueName: string,
					expectedScriptName: string,
					expectedEnvName?: string
				) {
					const requests = { count: 0 };
					let resource = `accounts/:accountId/workers/queues/:expectedQueueName/consumers/:expectedScriptName`;
					if (expectedEnvName !== undefined) {
						resource += `/environments/:expectedEnvName`;
					}
					msw.use(
						rest.delete(`*/${resource}`, async (request, response, context) => {
							requests.count++;
							expect(request.params.accountId).toBe("some-account-id");
							expect(request.params.expectedQueueName).toBe(expectedQueueName);
							expect(request.params.expectedScriptName).toBe(
								expectedScriptName
							);
							return response.once(
								context.status(200),
								context.json({
									success: true,
									errors: [],
									messages: [],
									result: {},
								})
							);
						})
					);

					return requests;
				}

				it("should show the correct help text", async () => {
					await runWrangler("queues consumer remove --help");
					expect(std.err).toMatchInlineSnapshot(`""`);
					expect(std.out).toMatchInlineSnapshot(`
				"wrangler queues consumer remove <queue-name> <script-name>

				Remove a Queue Consumer

				Positionals:
				  queue-name   Name of the queue to configure  [string] [required]
				  script-name  Name of the consumer script  [string] [required]

				Flags:
				  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
				  -c, --config                    Path to .toml configuration file  [string]
				  -e, --env                       Environment to use for operations and .env files  [string]
				  -h, --help                      Show help  [boolean]
				  -v, --version                   Show version number  [boolean]"
			`);
				});

				it("should delete a consumer with no --env", async () => {
					const requests = mockDeleteRequest("testQueue", "testScript");
					await runWrangler("queues consumer remove testQueue testScript");

					expect(requests.count).toEqual(1);
					expect(std.out).toMatchInlineSnapshot(`
						"Removing consumer from queue testQueue.
						Removed consumer from queue testQueue."
					`);
				});

				it("should delete a consumer with --env", async () => {
					const requests = mockDeleteRequest(
						"testQueue",
						"testScript",
						"myEnv"
					);
					await runWrangler(
						"queues consumer remove testQueue testScript --env myEnv"
					);

					expect(requests.count).toEqual(1);
					expect(std.out).toMatchInlineSnapshot(`
						"Removing consumer from queue testQueue.
						Removed consumer from queue testQueue."
					`);
				});
			});
		});
	});
});
