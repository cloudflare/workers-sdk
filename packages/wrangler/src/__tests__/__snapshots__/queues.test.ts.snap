// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wrangler > queues > create > wrangler.json > should create a queue 1`] = `
"ðŸŒ€ Creating queue 'testQueue'
âœ… Created queue 'testQueue'

Configure your Worker to send messages to this queue:

{
  \\"queues\\": {
    \\"producers\\": [
      {
        \\"queue\\": \\"testQueue\\",
        \\"binding\\": \\"testQueue\\"
      }
    ]
  }
}
Configure your Worker to consume messages from this queue:

{
  \\"queues\\": {
    \\"consumers\\": [
      {
        \\"queue\\": \\"testQueue\\"
      }
    ]
  }
}"
`;

exports[`wrangler > queues > create > wrangler.json > should send queue settings with delivery delay 1`] = `
"ðŸŒ€ Creating queue 'testQueue'
âœ… Created queue 'testQueue'

Configure your Worker to send messages to this queue:

{
  \\"queues\\": {
    \\"producers\\": [
      {
        \\"queue\\": \\"testQueue\\",
        \\"binding\\": \\"testQueue\\"
      }
    ]
  }
}
Configure your Worker to consume messages from this queue:

{
  \\"queues\\": {
    \\"consumers\\": [
      {
        \\"queue\\": \\"testQueue\\"
      }
    ]
  }
}"
`;

exports[`wrangler > queues > create > wrangler.toml > should create a queue 1`] = `
"ðŸŒ€ Creating queue 'testQueue'
âœ… Created queue 'testQueue'

Configure your Worker to send messages to this queue:

[[queues.producers]]
queue = \\"testQueue\\"
binding = \\"testQueue\\"

Configure your Worker to consume messages from this queue:

[[queues.consumers]]
queue = \\"testQueue\\"
"
`;

exports[`wrangler > queues > create > wrangler.toml > should send queue settings with delivery delay 1`] = `
"ðŸŒ€ Creating queue 'testQueue'
âœ… Created queue 'testQueue'

Configure your Worker to send messages to this queue:

[[queues.producers]]
queue = \\"testQueue\\"
binding = \\"testQueue\\"

Configure your Worker to consume messages from this queue:

[[queues.consumers]]
queue = \\"testQueue\\"
"
`;
